src/backend/distributed/planner/recursive_planning.c

```diff
@@ -599,6 +611,267 @@ RecursivelyPlanNonColocatedSubqueriesInWhere(Query *query,
+														  IsDistributedTableRTE);
+}
+
+
+/*
+ * RecursivelyPlanJoin performs recursive planning to enable outer joins
+ * of the form <recurring> LEFT JOIN <distributed>, as well as RIGHT JOIN
+ * and FULL JOIN equivalents, by recursively planning the distributed
+ * side of the join.
+ *
+ * We also optimize the query plan by performing an inner join in the
+ * subquery.
+ */
+static void
+RecursivelyPlanJoin(Node *recurringNode, JoinExpr *joinExpr, Node *distributedNode,
+					Query *currentQuery,
+					RecursivePlanningContext *recursivePlanningContext)
+{
+	if (IsA(distributedNode, JoinExpr))
+	{
+		/*
+		 * TODO: support recursively planning nested joins
+		 *
+		 * Cases likes:
+		 *   <recurring> LEFT JOIN (<distributed> join <distributed>)
+		 *
+		 * would require moving part of the join tree into a subquery.
+		 * However, that implies we need to rebuild the rtable and repoint
+		 * all the Vars to the new rtable indexes. We have not implemented
+		 * that yet.
+		 */
+		return;
+	}
+
+	Assert(IsA(recurringNode, RangeTblRef));
+	Assert(IsA(distributedNode, RangeTblRef));
+
+	RangeTblRef *distributedRef = (RangeTblRef *) distributedNode;
+	RangeTblEntry *distributedRte = rt_fetch(distributedRef->rtindex,
+											 currentQuery->rtable);
+
+	if (distributedRte->rtekind == RTE_RELATION)
+	{
+		/* <recurring> LEFT JOIN <distributed table> */
+
+		PlannerRestrictionContext *restrictionContext =
+			GetPlannerRestrictionContext(recursivePlanningContext);
+		List *requiredAttributes =
+			RequiredAttrNumbersForRelation(distributedRte, restrictionContext);
+
+		TransformRelationRTEIntoSubqueryRTE(distributedRte, requiredAttributes,
+											recursivePlanningContext);
+	}
+	else if (distributedRte->rtekind != RTE_SUBQUERY)
+	{
+		/*
+		 * Ignore other RTEs for now, e.g. a function RTE could still be
+		 * distributed if one of its arguments is a sublink.
+		 */
+		return;
+	}
+
+	/*
+	 * If the recurring node is a single RTE, we optimize the query plan
+	 * by first doing in inner join within the subquery. That way, we
+	 * can typically reduce the result set size before applying the
+	 * outer join between the recurring node and the subquery result.
+	 *
+	 * If the recurring node is a join tree, we only recursively plan
+	 * the distributed node to avoid the complications of building a
+	 * new rtable.
+	 */
+	if (IsA(recurringNode, RangeTblRef))
+	{
+		RangeTblRef *recurringRef = (RangeTblRef *) recurringNode;
+		RangeTblEntry *recurringRte = rt_fetch(recurringRef->rtindex,
+											   currentQuery->rtable);
+
+		distributedRte->subquery =
+			WrapSubqueryInInnerJoin(recurringRte, joinExpr, distributedRte);
+	}
+
+	RecursivelyPlanSubquery(distributedRte->subquery, recursivePlanningContext);
+}
+
+
+/*
+ * WrapSubqueryInInnerJoin creates a new query that joins the recurringNode
+ * with the subquery.
+ */
+static Query *
+WrapSubqueryInInnerJoin(RangeTblEntry *recurringRte, JoinExpr *joinExpr,
```

Comment:
This seems to be not the solution we are searching for. It solves the problem for Non-match rows, but it returns duplicate rows for matching rows. I think we might need a little more sophisticated approach than this one. 

The following is a case where we can show that a simple inner join in not enough


```SQL
truncate ref, dist ;

insert into ref VALUES (1), (1);
insert into dist VALUES (1), (2);

-- on local tables
select * from ref LEFT JOIN dist ON (dist.a = ref.a);
┌───┬───┐
│ a │ a │
├───┼───┤
│ 1 │ 1 │
│ 1 │ 1 │
└───┴───┘
(2 rows)
-- on dist tables
select * from ref LEFT JOIN (SELECT * from ref JOIN dist USING (a) OFFSET 0) as foo USING(a);
┌───┐
│ a │
├───┤
│ 1 │
│ 1 │
│ 1 │
│ 1 │
└───┘
(4 rows)
```